### PR TITLE
feat(AWS IAM): Group IAM-related settings under `provider.iam`

### PR DESCRIPTION
**Pull request by runk**
Closes: #8681

This change groups all IAM-related properties of `provider` under `iam` property like so:

new way:

```yaml
provider:
  iam:
    role: # can be either an arn string or an object like below:
      statements:
        - Effect: 'Allow',
          Resource: '*',
          NotAction: 'iam:DeleteUser',
      managedPolicies:
        - 'arn:aws:iam::123456789012:user/*',
      permissionBoundary: arn:aws:iam::123456789012:policy/boundaries
    deploymentRole: arn:aws:iam::123456789012:role/my-role
```

old way

```yaml
provider:
    role: arn:aws:iam::123456789012:role/my-role
    iamRoleStatements:
        - Effect: 'Allow',
          Resource: '*',
          NotAction: 'iam:DeleteUser',
     iamManagedPolicies:
        - 'arn:aws:iam::123456789012:user/*',
    rolePermissionsBoundary: arn:aws:iam::123456789012:policy/boundaries
    cfnRole: arn:aws:iam::123456789012:role/my-role
```

### CODE REVIEW

**Comment by runk** [comment]
@pgrzesik @medikoo can you please advise on the approach and whether you have any thoughts on it? Want to make sure we're on the same page before I change docs & clean up code a bit.

File: lib/plugins/aws/utils/iam.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
+'use strict';
+
+exports.getDeploymentRole = provider => {
+  if (provider.iam && provider.iam.deploymentRole) {
+    return provider.iam.deploymentRole;
+  }
+
+  if (provider.cfnRole) {
+    // Warn?
```
Comment (pgrzesik) [inline lib/plugins/aws/utils/iam.js:9]: Here we should log deprecations as outlined in: https://github.com/serverless/serverless/blob/master/CONTRIBUTING.md#configuring-deprecations

File: docs/deprecations.md [hunk 2]
```diff
@@ -6,6 +6,12 @@ layout: Doc
 
 # Serverless Framework Deprecations
 
+<a name="PROVIDER_IAM_SETTINGS"><div>&nbsp;</div></a>
+
+## Grouping IAM settings under `provider.iam`
```
Comment (pgrzesik) [inline docs/deprecations.md:11]: Here it would be good to include information that the old syntax will be deprecated starting with `v3.0.0` version :+1:

File: test/unit/lib/plugins/aws/deployFunction.test.js [hunk 3]
```diff
@@ -417,7 +417,47 @@ describe('AwsDeployFunction', () => {
           },
           Timeout: 12,
           MemorySize: 512,
-          Role: 'arn:aws:us-east-1:123456789012:role/role',
+          Role: 'my-role',
+        })
+      ).to.be.equal(true);
+    });
+
+    it('should inherit provider-level iam config', async () => {
```
Comment (pgrzesik) [inline test/unit/lib/plugins/aws/deployFunction.test.js:425]: All new tests should be run with the use of `runServerless` utility as outlined here: https://github.com/serverless/serverless/blob/master/CONTRIBUTING.md#configuring-deprecations

**Review by pgrzesik — COMMENTED** [review]
Hello @runk, thanks for submitting your changes. Overall you're on a great path, I didn't provide an extensive review as it's still Work In Progress, but I've added a few comments that I believe will be useful, I'm looking forward to reviewing the final version :raised_hands:

**Comment by runk** [comment]
@pgrzesik can you please check it again? 🙇

File: test/unit/lib/plugins/aws/deployFunction.test.js [hunk 4]
```diff
@@ -417,10 +417,19 @@ describe('AwsDeployFunction', () => {
           },
           Timeout: 12,
           MemorySize: 512,
-          Role: 'arn:aws:us-east-1:123456789012:role/role',
+          Role: 'my-role',
         })
       ).to.be.equal(true);
     });
+
+    it('should prefer provider.iam.role over provider.role config', async () => {
```
Comment (pgrzesik) [inline test/unit/lib/plugins/aws/deployFunction.test.js:425]: I know it adds a bit of extra work here, but I think this test case should be written using `runServerless` as we're trying to move from these type of tests altogether.

File: test/unit/lib/plugins/aws/package/lib/mergeIamTemplates.test.js [hunk 5]
```diff
@@ -245,6 +263,108 @@ describe('lib/plugins/aws/package/lib/mergeIamTemplates.test.js', () => {
...
+            [
+              'arn:',
+              { Ref: 'AWS::Partition' },
+              ':iam::aws:policy/service-role/AWSLambdaVPCAccessExecutionRole',
+            ],
+          ],
+        });
+      });
+
+      it('should support `provider.logRetentionInDays`', () => {
```
Comment (pgrzesik) [inline test/unit/lib/plugins/aws/package/lib/mergeIamTemplates.test.js:359]: Is there a reason this was included as a part of this PR or it's a test unrelated to main changes?

**Review by pgrzesik — CHANGES_REQUESTED** [review]
Looks good overall, I added a few comments, you might also want to rebase on top of master to resolve  conflicts. I also added review request for @medikoo

File: test/unit/lib/plugins/aws/deployFunction.test.js (hunk 4 above)
Comment (runk) [reply test/unit/lib/plugins/aws/deployFunction.test.js:425]: I tried to find an example, however couldn't find one. I assumed it's due to the fact it's about deployment and stuff? Happy to change though

File: test/unit/lib/plugins/aws/package/lib/mergeIamTemplates.test.js (hunk 5 above)
Comment (runk) [reply test/unit/lib/plugins/aws/package/lib/mergeIamTemplates.test.js]: Ops.

File: lib/plugins/aws/provider.js [hunk 6]
```diff
@@ -454,7 +480,7 @@ class AwsProvider {
           },
           awsLambdaTimeout: { type: 'integer', minimum: 1, maximum: 900 },
           awsLambdaTracing: { anyOf: [{ enum: ['Active', 'PassThrough'] }, { type: 'boolean' }] },
-          awsLambdaVersionning: { type: 'boolean' },
+          awsLambdaVersioning: { type: 'boolean' },
```
[the remark this answers is not in the document]
Comment (medikoo) [reply lib/plugins/aws/provider.js:483]: Can we propose it with other PR? It's a bug fix that deserves other individual commit in a history

File: lib/plugins/aws/provider.js [hunk 7]
```diff
@@ -356,6 +361,27 @@ class AwsProvider {
...
+              {
+                type: 'object',
+                properties: {
+                  managedPolicies: { $ref: '#/definitions/awsArnList' },
+                  statements: { $ref: '#/definitions/awsIamPolicyStatements' },
+                  permissionBoundary: { $ref: '#/definitions/awsArnString' },
+                },
+              },
+            ],
+          },
```
Comment (medikoo) [inline lib/plugins/aws/provider.js:384]: Definitions are intended for reusable definitions. If I understand correctly both of above address specific single case. In such scenario let's put definitions directly at their schema endpoints

File: lib/plugins/aws/provider.js [hunk 8]
```diff
@@ -285,6 +286,10 @@ class AwsProvider {
             type: 'string',
             pattern: '^arn:',
           },
+          awsArnList: {
+            type: 'array',
+            items: { $ref: '#/definitions/awsArnString' },
+          },
```
Comment (medikoo) [inline lib/plugins/aws/provider.js:292]: Let's not introduce such definition. I think we may find tens of other cases where we could go in such direction, while it quickly pollutes the definitions with noise. 

Also seeing right at the point that we deal with an _array_ and not a singular value, has a value.

File: lib/plugins/aws/utils/iam.js [hunk 9]
```diff
@@ -0,0 +1,33 @@
+'use strict';
+
+exports.getDeploymentRole = (serverless) => {
```
[the remark this answers is not in the document]
Comment (medikoo) [reply lib/plugins/aws/utils/iam.js:3]: I believe right way to approach that is to:

1. Introduce `provider.customDeploymentRole` (where `provider` is instance of `AwsProvider` and update all `service.provider.cfnRole` occurences to rely on `provider.customDeploymentRole` instead

2. Introduce `provider.customExecutionRole` and update all points where we access it, so this property is accessed.

Also this is a refactor that we can do in other PR, and then in this PR just update initial resolution of `provider.customDeploymentRole` and `provider.customExecutionRole`.

It'll be great to split this work into multiple PR's that way. There's less chance risk we break something and history of changes in repo is much better (note that one PR ends as single commit)

File: lib/plugins/aws/package/lib/mergeIamTemplates.js [hunk 10]
```diff
@@ -1,14 +1,13 @@
...
-const BbPromise = require('bluebird');
 const path = require('path');
 
 module.exports = {
-  mergeIamTemplates() {
+  async mergeIamTemplates() {
     // resolve early if no functions are provided
     if (!this.serverless.service.getAllFunctions().length) {
-      return BbPromise.resolve();
+      return;
```
Comment (medikoo) [inline lib/plugins/aws/package/lib/mergeIamTemplates.js:10]: This function should be simply refactored to be a _sync_ function, but all such refactors best if are done separetely in context of other dedicated PR's

File: lib/plugins/aws/package/lib/mergeIamTemplates.js [hunk 11]
```diff
@@ -38,19 +37,30 @@ module.exports = {
...
+        'PROVIDER_IAM_SETTINGS',
+        'Starting with version 3.0.0, "provider.role" property will be replaced with "provider.iam.role"'
+      );
+      return;
     }
 
     // resolve early if all functions contain a custom role
-    const customRolesProvided = [];
-    this.serverless.service.getAllFunctions().forEach((functionName) => {
+    const customRolesProvided = this.serverless.service.getAllFunctions().every((functionName) => {
```
Comment (medikoo) [inline lib/plugins/aws/package/lib/mergeIamTemplates.js:57]: Let's not mix plain refactors into this PR, let's propose them with other PR (we can merge them right away)

This PR touches sensitive part and best if we only propose the code that is specific to proposed change

File: test/unit/lib/plugins/aws/deployFunction.test.js (hunk 4 above)
Comment (medikoo) [reply test/unit/lib/plugins/aws/deployFunction.test.js:425]: @runk all testing related documentation can be found here: https://github.com/serverless/serverless/tree/master/test#how-to-run-organize-and-write-tests

File: test/unit/lib/plugins/aws/package/lib/mergeIamTemplates.test.js [hunk 12]
```diff
@@ -245,6 +263,98 @@ describe('lib/plugins/aws/package/lib/mergeIamTemplates.test.js', () => {
       });
     });
 
+    describe('Provider properties - modern usage', () => {
```
Comment (medikoo) [inline test/unit/lib/plugins/aws/package/lib/mergeIamTemplates.test.js:266]: We should rather update `Provider properties` to use just new versions, and introduce `Provider deprecated properties` to test those deprecated.

It'll also open the door for very easy update when we will upgrade the major (we will simply remove `Provider deprecated properties` section)

**Review by medikoo — CHANGES_REQUESTED** [review]
Thank you @runk, that's great outstanding work!

Please see my comments, as many independent areas are touched I believe it should be split into few PR's

File: lib/plugins/aws/utils/iam.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (runk) [reply lib/plugins/aws/utils/iam.js:3]: @medikoo makes sense re deployment role.

Not 100% sure how you think it should work for execution role when function params are in the mix?

[the remark this answers is not in the document]
Comment (medikoo) [reply lib/plugins/aws/utils/iam.js:3]: > Not 100% sure how you think it should work for execution role when function params are in the mix?

Good point, I forgot it can be customized per function.

In light of that I propose to introduce `provider.resolveFunctionIamRoleArn` method (we already have e.g. [`  resolveFunctionIamRoleResourceName`](https://github.com/serverless/serverless/blob/2b7568a960c88dda8ab2bbe1b6c8dd238fa78a51/lib/plugins/aws/provider.js#L1537))

File: test/unit/lib/plugins/aws/deploy/lib/createStack.test.js [hunk 13]
```diff
@@ -88,17 +88,27 @@ describe('createStack', () => {
...
-      return awsDeploy.create().then(() => {
-        expect(createStackStub.args[0][2].RoleARN).to.equal(
-          'arn:aws:iam::123456789012:role/myrole'
-        );
-      });
+      await awsDeploy.create();
+      expect(createStackStub.args[0][2].RoleARN).to.equal('arn:aws:iam::123456789012:role/role-b');
+    });
+
+    it('should use iam.deploymentRole service role if set', async () => {
```
Comment (runk) [inline test/unit/lib/plugins/aws/deploy/lib/createStack.test.js:101]: I know all new tests should be using `runServerless`, however with `deploy` cli arg it seems to run for unreasonable 5-10 seconds (as in deploy/index.test.js). Not sure I should doing it or keeping this simple and consistent with the rest?

Another con of using `runServerless` for deploy here is massive map of stubs similar to https://github.com/serverless/serverless/blob/master/test/unit/lib/plugins/aws/deploy/index.test.js#L225-L281

That said I'm still happy to do it if you reckon it's worthwhile and related to the current change.

**Comment by runk** [comment]
@medikoo @pgrzesik can you please check it again? 🙇 

With most (or all?) unrelated changes done in separate PRs hoping it'll be easier to review.

File: lib/plugins/aws/provider.js [hunk 14]
```diff
@@ -768,6 +768,29 @@ class AwsProvider {
...
             },
+            iam: {
+              type: 'object',
+              properties: {
+                role: {
+                  anyOf: [
+                    { $ref: '#/definitions/awsArnString' },
+                    {
+                      type: 'object',
+                      properties: {
```
Comment (pgrzesik) [inline lib/plugins/aws/provider.js:779]: Should we also add `additionalProperties: false` for that definition?

File: test/unit/lib/plugins/aws/deploy/lib/createStack.test.js (hunk 13 above)
Comment (pgrzesik) [reply test/unit/lib/plugins/aws/deploy/lib/createStack.test.js:101]: Let's not extend the old tests in any way like here. We should be using `runServerless` for new tests. When testing e.g. `deploy`, you can use `lastLifecycleHookName`: https://github.com/serverless/test/blob/master/docs/run-serverless.md#lastlifecyclehookname to avoid running hooks that are not essential for your test. I know some of these runs might be longer, but the structure of current tests here is really fragile and makes it harder to refactor functionality which is why we prefer to use more of an "integration-based" approach to testing

**Review by pgrzesik — CHANGES_REQUESTED** [review]
Thank you @runk, it looks good overall, I just have a few small comments, please let me know

File: test/unit/lib/plugins/aws/deploy/lib/createStack.test.js (hunk 13 above)
Comment (runk) [reply test/unit/lib/plugins/aws/deploy/lib/createStack.test.js:140]: Converted tests to use runServerless however not sure what hook I should be using to make them faster than 10s

**Comment by runk** [comment]
@pgrzesik check again?

File: test/unit/lib/plugins/aws/deploy/lib/createStack.test.js (hunk 13 above)
Comment (medikoo) [reply test/unit/lib/plugins/aws/deploy/lib/createStack.test.js:140]: I remember in some situations those runs being slow due to custom resources being involved (not sure if here it's the case). Still then fix was to simply stub `spawn` via [`shouldStubSpawn`](https://github.com/serverless/test/blob/master/docs/run-serverless.md#shouldstubspawn) option

Comment (pgrzesik) [reply test/unit/lib/plugins/aws/deploy/lib/createStack.test.js:140]: Thanks @runk for going the `runServerless` route. It would also be great to move these `runServerless`-based tests to a separate `describe` as mentioned in: https://github.com/serverless/serverless/blob/master/test/README.md It allows to easier asses which tests are using `runServerless` and how many of them are still using the "old way". Let's name the `describe` block after the name of this test file.

File: test/unit/lib/plugins/aws/deploy/lib/createStack.test.js [hunk 15]
```diff
@@ -88,17 +137,47 @@ describe('createStack', () => {
...
+          },
+        },
+        awsRequestStubMap: makeStubMap(createStackStub),
+        lastLifecycleHookName: 'deploy:function:deploy',
+      });
+      expect(createStackStub).to.be.calledOnce;
+      expect(createStackStub.getCall(0).firstArg.RoleARN).to.equal(
+        'arn:aws:iam::123456789012:role/role-a'
+      );
+      createStackStub.reset();
```
Comment (pgrzesik) [inline test/unit/lib/plugins/aws/deploy/lib/createStack.test.js:160]: Maybe we could initialize that stub outside of this specific test and reset it's history (and behavior if needed) in `beforeEach` or `afterEach`?

File: test/unit/lib/plugins/aws/deploy/lib/createStack.test.js (hunk 13 above)
Comment (runk) [reply test/unit/lib/plugins/aws/deploy/lib/createStack.test.js]: Moved tests into separate describe. Trick with `shouldStubSpawn: true` didn't help with long test time though..

File: test/unit/lib/plugins/aws/deploy/lib/createStack.test.js (hunk 15 above)
Comment (runk) [reply test/unit/lib/plugins/aws/deploy/lib/createStack.test.js]: Yeah, tried it. Not sure why, however behaviour doesn't seem being reset when same stub map used for the second time. Not sure what's going on tbh..

Comment (pgrzesik) [reply test/unit/lib/plugins/aws/deploy/lib/createStack.test.js]: I think we don't have to reset behavior as it's staying the same across two tests and the only thing that we should reset is history via `createStackStub.resetHistory()` - you can see a working example here: https://github.com/serverless/serverless/blob/7c998e03ca104948c6c2673c9fee6992689219df/test/unit/lib/plugins/aws/provider.test.js#L1850

**Comment by pgrzesik** [comment]
Thank you @runk - it looks good overall, I added a minor comment about reusing stubs across tests, please let me know if that will work here. 

@medikoo would you be able to review this once again before wrapping up this PR?

File: test/unit/lib/plugins/aws/deploy/lib/createStack.test.js (hunk 13 above)
Comment (pgrzesik) [reply test/unit/lib/plugins/aws/deploy/lib/createStack.test.js]: Yeah, in this case that might not actually do the trick. I believe we need to investigate these tests separately and evaluate what's causing them to run so long, as we already have a few similar ones with longer running times.

File: test/unit/lib/plugins/aws/deploy/lib/createStack.test.js (hunk 15 above)
Comment (runk) [reply test/unit/lib/plugins/aws/deploy/lib/createStack.test.js]: It did the trick! I think the behaviour reset of `describeStacksStub` was doing different reset from what I was expecting - I wanted to start counting calls from scratch, but it was actually was removing all that perhaps.

On side topic I can't really explain why it has this behaviour - copied it from elsewhere, but this magic is needed 🤕

Comment (pgrzesik) [reply test/unit/lib/plugins/aws/deploy/lib/createStack.test.js]: It's due to the fact that `sinon` offers two different resets - you can either reset the behavior via `stub.resetBehavior` or calls history via `stub.resetHistory`. You can also call `stub.reset` to do both at the same time - in this case, the behavior should stay the same, hence only `resetHistory` is needed :+1:

File: lib/plugins/aws/package/lib/mergeIamTemplates.js [hunk 16]
```diff
@@ -37,7 +37,14 @@ module.exports = {
         );
       });
 
+    const { iam } = this.serverless.service.provider;
+    const iamRole = iam && iam.role ? iam.role : {};
```
Comment (medikoo) [inline lib/plugins/aws/package/lib/mergeIamTemplates.js:41]: We can put it simpler as:

```javascript
const iamRole = _.get(this.serverless.service.provider.iam, 'role', {})
```

File: lib/plugins/aws/package/lib/mergeIamTemplates.js [hunk 17]
```diff
@@ -37,7 +37,14 @@ module.exports = {
...
+
     // resolve early if provider level role is provided
+    if (typeof iamRole === 'string') {
+      return;
+    }
+
     if ('role' in this.serverless.service.provider) {
```
Comment (medikoo) [inline lib/plugins/aws/package/lib/mergeIamTemplates.js:48]: I think we should show deprecation here as well (?)

File: lib/plugins/aws/package/lib/mergeIamTemplates.js [hunk 18]
```diff
@@ -66,7 +73,14 @@ module.exports = {
...
-    if (this.serverless.service.provider.rolePermissionsBoundary) {
+    // set permission boundary
+    if (iamRole.permissionBoundary) {
+      iamRoleLambdaExecutionTemplate.Properties.PermissionsBoundary = iamRole.permissionBoundary;
+    } else if (this.serverless.service.provider.rolePermissionsBoundary) {
+      this.serverless._logDeprecation(
+        'PROVIDER_IAM_SETTINGS',
+        'Starting with version 3.0.0, "provider.rolePermissionsBoundary" property will be replaced with "provider.iam.role.permissionBoundary"'
+      );
       iamRoleLambdaExecutionTemplate.Properties.PermissionsBoundary = this.serverless.service.provider.rolePermissionsBoundary;
```
Comment (medikoo) [inline lib/plugins/aws/package/lib/mergeIamTemplates.js:84]: If users sets both `provider.iam.role.permissionBoundary` and `provider.rolePermissionsBoundary`, we should favor first (which we do now), but we still should show information that deprecated property was configured (It's important to do it, as introducing not supported properties with next version will crash the deployment)

File: lib/plugins/aws/package/lib/mergeIamTemplates.js [hunk 19]
```diff
@@ -66,7 +73,14 @@ module.exports = {
     iamRoleLambdaExecutionTemplate.Properties.Path = this.provider.naming.getRolePath();
     iamRoleLambdaExecutionTemplate.Properties.RoleName = this.provider.naming.getRoleName();
 
-    if (this.serverless.service.provider.rolePermissionsBoundary) {
+    // set permission boundary
+    if (iamRole.permissionBoundary) {
+      iamRoleLambdaExecutionTemplate.Properties.PermissionsBoundary = iamRole.permissionBoundary;
+    } else if (this.serverless.service.provider.rolePermissionsBoundary) {
+      this.serverless._logDeprecation(
```
Comment (medikoo) [inline lib/plugins/aws/package/lib/mergeIamTemplates.js:80]: Deprecations should be configured at `initialize` hook (as e.g. here: https://github.com/serverless/serverless/blob/12805c3d152d85af9dba3dd3ecfa2002a621f6a8/lib/plugins/aws/package/compile/functions.js#L31-L61), as that guarantees that deprecation are shown at early stage of command processing and even if affected functionality is not used (e.g. we do not package service, or via some other logic processing of those properties is skipped).

It's important as with next major we will throw on not supported properties, so deprecation of existing properties needs to be communicated in case of all commands.

File: test/unit/lib/plugins/aws/package/lib/mergeIamTemplates.test.js [hunk 20]
```diff
@@ -245,6 +263,98 @@ describe('lib/plugins/aws/package/lib/mergeIamTemplates.test.js', () => {
...
+        const IamRoleLambdaExecution = naming.getRoleLogicalId();
+        const {
+          Properties: { PermissionsBoundary },
+        } = cfResources[IamRoleLambdaExecution];
+        expect(PermissionsBoundary).to.be.equal(
+          'arn:aws:iam::123456789012:policy/XCompanyBoundaries'
+        );
+      });
+
+      it('should ensure needed IAM configuration when `provider.vpc` is configured', () => {
```
Comment (medikoo) [inline test/unit/lib/plugins/aws/package/lib/mergeIamTemplates.test.js:341]: Let's remove that test from `Provider properties - deprecated properties` block (at least to my understanding this is not influenced by properties we were handling here, so we're now testing twice same thing)

**Review by medikoo — CHANGES_REQUESTED** [review]
@runk looks very good. I've just realized one issue with point at which we report deprecations. Sorry for not raising this earlier.

Otherwise all looks great to me.

File: test/unit/lib/plugins/aws/deploy/lib/createStack.test.js (hunk 13 above)
Comment (pgrzesik) [reply test/unit/lib/plugins/aws/deploy/lib/createStack.test.js]: Just fyi: I believe I've managed to find the cause for long-running tests and fix is incoming: https://github.com/serverless/test/pull/75

File: lib/plugins/aws/package/lib/mergeIamTemplates.js (hunk 17 above)
Comment (runk) [reply lib/plugins/aws/package/lib/mergeIamTemplates.js:47]: Sure, I was expecting that all (or most?) usage will result in calling code from `provider.getCustomExecutionRole()` that does report the deprecation.

File: lib/plugins/aws/provider.js [hunk 21]
```diff
@@ -1806,6 +1846,17 @@ class AwsProvider {
       );
     }
   }
+
+  reportDeprecatedUsage(deprecationCode, pairs) {
```
Comment (runk) [inline lib/plugins/aws/provider.js:1850]: Let me know if I should test this function, and how :) Making it standalone would make it easier.

**Comment by runk** [comment]
Hey @medikoo and @pgrzesik 👋 

I believe I addresses the last round of feedback - please check again

File: lib/plugins/aws/provider.js [hunk 22]
```diff
@@ -1806,6 +1846,17 @@ class AwsProvider {
...
+    for (const [deprecated, modern] of pairs) {
+      if (_.has(this.serverless.service.provider, deprecated)) {
```
Comment (medikoo) [inline lib/plugins/aws/provider.js:1852]: Let's not use `_.has` (we can do `this.serverless.service.provider[deprecated] != null`)

File: lib/plugins/aws/provider.js [hunk 23]
```diff
@@ -1806,6 +1846,17 @@ class AwsProvider {
...
+        this.serverless._logDeprecation(
```
Comment (medikoo) [inline lib/plugins/aws/provider.js:1853]: Deprecation logger works in a way that it reports given deprecation (code) only once.

So with the way it's constructed now, all iteration after first log, is ineffective.

Best way would be to first gather info of all deprecated properties, and then list all of them with one log.

File: lib/plugins/aws/provider.js (hunk 21 above)
Comment (medikoo) [reply lib/plugins/aws/provider.js:1850]: We can put it into `lib/utils/report-deprecated-properties.js`

I'd also use object or map as an input (not array). It may work that way, and under the hood use [`utils/logDeprecation.js`](https://github.com/serverless/serverless/blob/816394c6e5dfc50b332314aef66eeb9ed75d139a/lib/utils/logDeprecation.js)

```javascript
reportDeprecatedProperties('PROVIDER_IAM_SETTINGS', {
  role: 'iam.role',
  ...
}, { serviceConfig: this.serverless.service });
```

**Review by medikoo — CHANGES_REQUESTED** [review]
Thank you @runk looks great! Please see my comments

File: lib/plugins/aws/provider.js (hunk 23 above)
Comment (runk) [reply lib/plugins/aws/provider.js:1853]: Not sure it's fair a comparison, as there's still going to be a loop, and still at least one call to `this.serverless._logDeprecation`. Up to you really 🤷‍♂️

File: lib/plugins/aws/provider.js (hunk 22 above)
Comment (runk) [reply lib/plugins/aws/provider.js:1852]: If you want to make it an util, I'd make props like `provider.role` etc so it's also usable for deprecating function props and so on. This will require usage of `_.get` or `_.has` Thoughts?

File: lib/plugins/aws/provider.js (hunk 21 above)
Comment (runk) [reply lib/plugins/aws/provider.js:1850]: I can see file name casing is all over the place, should I use kebab case or camel case?

Comment (medikoo) [reply lib/plugins/aws/provider.js:1850]: We plan to rename all to kebab case. So you can introduce new util in kebab case

File: lib/plugins/aws/provider.js (hunk 22 above)
Comment (medikoo) [reply lib/plugins/aws/provider.js:1852]: I think indeed that `_.has` is ok in that case

File: lib/plugins/aws/provider.js (hunk 23 above)
Comment (medikoo) [reply lib/plugins/aws/provider.js:1853]: Point, is to inform about all deprecated properties with single `logDeprecation` call (yes, we still need to for loop, that doesn't contradict that)

**Comment by runk** [comment]
@medikoo let me know if it's something along the line you were thinking. Once confirmed I'll add tests for the new util.

File: lib/utils/report-deprecated-properties.js [hunk 24]
```diff
@@ -0,0 +1,20 @@
...
+const _ = require('lodash');
+const logDeprecation = require('./logDeprecation');
+
+module.exports = (code, props, { serviceConfig } = {}) => {
+  const warnings = [];
+
+  for (const [oldProp, newProp] of Object.entries(props)) {
+    if (_.has(serviceConfig, oldProp)) {
+      warnings.push(
+        `Starting with version 3.0.0, "${oldProp}" property will be replaced with "${newProp}"`
```
Comment (medikoo) [inline lib/utils/report-deprecated-properties.js:12]: Thanks @runk for update. In general it looks good, I will just improve messaging, as when constructed as proposed itwill create a lot of repetitive noise (_"Starting with... Starting with... Starting with..."_)

Ideally if we simply group names of deprecated properties, and then construct one concise message as:

```
Starting with v3.0.0, support for "a", "b" and "c" will be removed. Please use "d", "e" and "f" instead
```

Comment (runk) [reply lib/utils/report-deprecated-properties.js]: Okay, I used slightly different formatting, please check (as well as tests)

cc @medikoo

**Review by medikoo — CHANGES_REQUESTED** [review]
Thank you @runk. It looks very good! I have just few final suggestions, and I believe we should be good to go

File: lib/utils/report-deprecated-properties.js [hunk 25]
```diff
@@ -0,0 +1,30 @@
...
+const logDeprecation = require('./logDeprecation');
+
+module.exports = (code, props, { serviceConfig } = {}) => {
+  const msg = module.exports.getMessage(props, serviceConfig);
+  if (msg) {
+    logDeprecation(code, msg, { serviceConfig });
+  }
+};
+
+module.exports.getMessage = (props, serviceConfig) => {
```
[the remark this answers is not in the document]
Comment (runk) [reply lib/utils/report-deprecated-properties.js:13]: This would work, however logDeprecation is messing around with line breaks/indentation and injecting colors. This makes this test reliant on business logic inside the `logDeprecation`. 

```
"Serverless: \u001b[38;2;255;165;0mDeprecation warning: Starting with version 3.0.0, following properties will be replaced:\u001b[39m\n\u001b[38;2;255;165;0m              \"provider.role\" -> \"provider.iam.role\"\u001b[39m\n\u001b[38;2;255;165;0m              \"provider.iamRoleStatements\" -> \"provider.iam.role.statements\"\u001b[39m\n            \u001b[38;2;255;165;0mMore Info: https://www.serverless.com/framework/docs/deprecations/#MY_CODE\u001b[39m\n"
```

vs 

```
Starting with version 3.0.0, following properties will be replaced:\n
  "provider.role" -> "provider.iam.role"\n
  "provider.iamRoleStatements" -> "provider.iam.role.statements"
```

I can make this work, however not sure how it makes it any better.

File: test/unit/lib/utils/report-deprecated-properties.test.js [hunk 26]
```diff
@@ -0,0 +1,85 @@
...
+            {
+              'provider.role': 'provider.iam.role',
+            },
+            { serviceConfig }
+          )
+      );
+      expect(stdoutData).to.include('Deprecation warning');
+      expect(stdoutData).to.include(
+        'https://www.serverless.com/framework/docs/deprecations/#MY_CODE'
+      );
```
[the remark this answers is not in the document]
Comment (runk) [reply test/unit/lib/utils/report-deprecated-properties.test.js:35]: If you want to use stdout in tests, why should I tap into `triggeredDeprecations`?

[the remark this answers is not in the document]
Comment (medikoo) [reply test/unit/lib/utils/report-deprecated-properties.test.js:35]: > If you want to use stdout in tests, why should I tap into triggeredDeprecations?

It's more that confirming on messages dedicated to humans is maintanance wise problematic. At any time we may decide to improve wording, grammar etc. and that should not break this test. 

Technically if we have a good programmatical mean to confirm on something, let's rely on it. In this case we have `triggeredDeprecations` collection

File: lib/utils/report-deprecated-properties.js (hunk 25 above)
[the remark this answers is not in the document]
Comment (medikoo) [reply lib/utils/report-deprecated-properties.js:13]: I think confirming just on some bits as `"provider.role" -> "provider.iam.role"` will still work. So I would do just that (let's do just sanity check that message went through, but not necessarily confirm on all its characters)

[the remark this answers is not in the document]
Comment (runk) [reply lib/utils/report-deprecated-properties.js:13]: It'll work, but the test by itself will be covering only parts of what code is going. Up to you really, but I wouldn't do it this way. 

I see no harm in exposing a fn if it makes testing very specific and isolated from upstream dependencies (aka low maintenance).

[the remark this answers is not in the document]
Comment (medikoo) [reply lib/utils/report-deprecated-properties.js:13]: > It'll work, but the test will be covering only parts of what the code is doing

In my understanding we still can maintain 100% coverage (if we want to), and confirming on every character on string, doesn't really expand the coverage, it just enforces us to put same not simple string in two different place (which maintenance wise is problematic)

[the remark this answers is not in the document]
Comment (runk) [reply lib/utils/report-deprecated-properties.js]: See again please @medikoo

**Review by medikoo — CHANGES_REQUESTED** [review]
@runk great thanks for update! Still I see that tests started to fail, do you know the cause?

**Comment by runk** [comment]
Yup, all fixed now @medikoo

**Review by medikoo — APPROVED** [review]
Looks great @runk, thank you!

I'm leaving it for final approval by @pgrzesik

**Review by pgrzesik — APPROVED** [review]
Thank you @runk for your tremendous work :clap: Big thanks to you @medikoo as well for providing a lot of great fedback during reviews. It's great to have that refactoring in and the introduced utility for reporting deprecated properties will also definitely be useful in the future in multiple cases :+1:

**Comment by G-Rath** [comment]
@pgrzesik @medikoo do you think it'd make sense to do this for API Gateway settings too?

(Asking here initially as I can't find existing issues about it but it's something that makes sense to me so want to check it's not already been considered/has an issue open before I make a new issue).

I'd be happy to help with the PR for that if so :)

**Comment by pgrzesik** [comment]
Hey @G-Rath - thank you - we're always open for enhancement proposals like that :100: The question I have here - what properties do you think are currently not grouped under `provider.apiGateway` that should be? We might also have that conversation in a separate issue :+1:

**Comment by G-Rath** [comment]
Some of the properties I had in mind were `endpointType` and `apiName`.

tbh looking over the list there are less properties than I thought 😅

> We might also have that conversation in a separate issue 👍

For sure!

**Comment by pgrzesik** [comment]
@G-Rath - would you be willing to start a new issue with the proposal for that refactoring? Thanks in advance :bow:
